crates/sui-indexer-alt-framework/src/postgres.rs

```diff
@@ -0,0 +1,278 @@
+    ///
+    /// Optional `migrations` contains the SQL to run in order to bring the database schema up-to-date for
+    /// the specific instance of the indexer, generated using diesel's `embed_migrations!` macro.
+    /// These migrations will be run as part of initializing the indexer if provided.
+    ///
+    /// After initialization, at least one pipeline must be added using [Self::concurrent_pipeline]
+    /// or [Self::sequential_pipeline], before the indexer is started using [Self::run].
+    pub async fn new_from_pg(
+        database_url: Url,
+        db_args: DbArgs,
+        indexer_args: IndexerArgs,
+        client_args: ClientArgs,
+        ingestion_config: IngestionConfig,
+        migrations: Option<&'static EmbeddedMigrations>,
+        registry: &Registry,
+        cancel: CancellationToken,
+    ) -> Result<Self> {
+        let db = Db::for_write(database_url, db_args) // I guess our store needs a constructor fn
+            .await
+            .context("Failed to connect to database")?;
+
+        // At indexer initialization, we ensure that the DB schema is up-to-date.
+        db.run_migrations(Self::migrations(migrations))
+            .await
+            .context("Failed to run pending migrations")?;
+
+        registry.register(Box::new(DbConnectionStatsCollector::new(
+            Some("indexer_db"),
+            db.clone(),
+        )))?;
+
+        Indexer::<Db>::new(
+            db,
+            indexer_args,
+            client_args,
+            ingestion_config,
+            registry,
+            cancel,
+        )
+        .await
+    }
+
+    /// Create a new temporary database and runs provided migrations in tandem with the migrations
+    /// necessary to support watermark operations on the indexer. The indexer is then instantiated
+    /// and returned along with the temporary database.
+    pub async fn new_for_testing(migrations: &'static EmbeddedMigrations) -> (Indexer<Db>, TempDb) {
+        let temp_db = TempDb::new().unwrap();
+        let store = Db::for_write(temp_db.database().url().clone(), DbArgs::default())
+            .await
+            .unwrap();
+        store
+            .run_migrations(Indexer::<Db>::migrations(Some(migrations)))
+            .await
+            .unwrap();
+
+        let indexer = Indexer::new(
+            store,
+            IndexerArgs::default(),
+            ClientArgs {
+                remote_store_url: None,
+                local_ingestion_path: Some(tempdir().unwrap().into_path()),
+                rpc_api_url: None,
+                rpc_username: None,
+                rpc_password: None,
+            },
+            IngestionConfig::default(),
+            &Registry::new(),
+            CancellationToken::new(),
+        )
+        .await
+        .unwrap();
+        (indexer, temp_db)
+    }
+
+    /// Returns new migrations derived from the combination of provided migrations and migrations as
```

Comment:
I don't think the implementation of this belongs here -- it should be part of `sui-pg-db`, because that's what carries the schema.